File: ios/PaintEngine/Sources/PaintEngine/Segment.swift
/// Connected regions + small-region merging — port of src/engine/segment.ts.
/// v3.1: CONTRAST-AWARE merging (CIEDE2000): a small region that stands out
/// against every neighbor (ΔE00 ≥ 12) is a detail (lips, eyelets, lace) and
/// survives; low-contrast small regions merge into the ΔE00-nearest neighbor.

public struct Region {
    public var id: Int
    public var colorIdx: Int
    public var area: Int
    public var pixels: [Int]
    public var minX: Int = 0
    public var minY: Int = 0
    public var maxX: Int = -1
    public var maxY: Int = -1
    public var neighbors: Set<Int> = []

    public init(id: Int, colorIdx: Int, pixels: [Int]) {
        self.id = id
        self.colorIdx = colorIdx
        self.area = pixels.count
        self.pixels = pixels
    }
}

public struct SegmentResult {
    /// Region id per pixel (compacted ids after merge).
    public var regionIds: [UInt32]
    public var regions: [Region]

    public init(regionIds: [UInt32], regions: [Region]) {
        self.regionIds = regionIds
        self.regions = regions
    }
}

/// Contrast threshold for the merge guard (ΔE00): small regions contrasting
/// MORE than this with every neighbor are details and are kept.
public let mergeContrastThreshold = 12.0

private let unassigned: UInt32 = 0xffffffff

/// Split a label map into 4-connected regions and record adjacency
/// from pixel pairs with differing labels (right/down pairs).
public func buildRegions(_ labels: [UInt16], _ width: Int, _ height: Int) -> SegmentResult {
    let n = width * height
    var regionIds = [UInt32](repeating: unassigned, count: n)
    var pixelsLists: [[Int]] = []
    var neighborSets: [Set<Int>] = []
    var queue = [Int](repeating: 0, count: n)

    for start in 0..<n {
        if regionIds[start] != unassigned { continue }
        let colorIdx = Int(labels[start])
        let id = pixelsLists.count
        var pixels: [Int] = []
        neighborSets.append([])
        var head = 0
        var tail = 0
        queue[tail] = start
        tail += 1
        regionIds[start] = UInt32(id)
        pixels.append(start)

        while head < tail {
            let p = queue[head]
            head += 1
            let x = p % width
            let y = p / width

            let q4 = [
                x + 1 < width ? p + 1 : -1,
                y + 1 < height ? p + width : -1,
                x > 0 ? p - 1 : -1,
                y > 0 ? p - width : -1,
            ]
            for q in q4 where q >= 0 {
                if Int(labels[q]) == colorIdx {
                    if regionIds[q] == unassigned {
                        regionIds[q] = UInt32(id)
                        queue[tail] = q
                        tail += 1
                        pixels.append(q)
                    }
                } else if regionIds[q] != unassigned {
                    let nb = Int(regionIds[q])
                    if nb != id {
                        neighborSets[id].insert(nb)
                        neighborSets[nb].insert(id)
                    }
                }
            }
        }
        pixelsLists.append(pixels)
    }

    var regions: [Region] = []
    regions.reserveCapacity(pixelsLists.count)
    for (id, pixels) in pixelsLists.enumerated() {
        var r = Region(id: id, colorIdx: Int(labels[pixels[0]]), pixels: pixels)
        r.neighbors = neighborSets[id]
        regions.append(r)
    }
    recomputeBboxes(&regions, regionIds, width, height)
    return SegmentResult(regionIds: regionIds, regions: regions)
}

private func recomputeBboxes(_ regions: inout [Region], _ regionIds: [UInt32], _ width: Int, _ height: Int) {
    for i in regions.indices {
        regions[i].minX = width
        regions[i].minY = height
        regions[i].maxX = -1
        regions[i].maxY = -1
    }
    for p in 0..<regionIds.count {
        let x = p % width
        let y = p / width
        let i = Int(regionIds[p])
        if x < regions[i].minX { regions[i].minX = x }
        if x > regions[i].maxX { regions[i].maxX = x }
        if y < regions[i].minY { regions[i].minY = y }
        if y > regions[i].maxY { regions[i].maxY = y }
    }
    _ = height
}

/// Merge every region smaller than minAreaPixels into its ΔE00-nearest
/// neighbor until none remain. Contrast-aware (v3.1): a small region whose
/// ΔE00 with EVERY alive neighbor is ≥ mergeContrastThreshold is kept —
/// details like lips/eyelets/lace survive. Isolated small regions are kept.
public func mergeSmallRegions(_ seg: inout SegmentResult, _ palette: [PaletteEntry], _ minAreaPixels: Int, _ width: Int, _ height: Int) {
    if minAreaPixels <= 0 { return }
    var regions = seg.regions
    var alive = [Bool](repeating: true, count: regions.count)
    var absorbedInto = [Int](repeating: -1, count: regions.count)

    var smallQueue: [Int] = []
    for r in regions where r.area < minAreaPixels { smallQueue.append(r.id) }

    let labOf: [Lab] = regions.map { palette[$0.colorIdx].lab }

    var qi = 0
    while qi < smallQueue.count {
        let smallId = smallQueue[qi]
        qi += 1
        if !alive[smallId] { continue }
        let smallArea = regions[smallId].area
        if smallArea >= minAreaPixels { continue }
        if regions[smallId].neighbors.isEmpty { continue } // isolated: keep

        // Contrast guard: a small region that stands out against EVERY alive
        // neighbor is a detail, not noise — keep it.
        var maxNeighborD = -1.0
        for nbId in regions[smallId].neighbors where alive[nbId] {
            let d = labDistance2000(labOf[smallId], labOf[nbId])
            if d > maxNeighborD { maxNeighborD = d }
        }
        if maxNeighborD < 0 { continue } // no alive neighbors: keep
        if maxNeighborD >= mergeContrastThreshold { continue } // contrast detail: keep

        // Nearest alive neighbor by ΔE00 among low-contrast candidates.
        var bestId = -1
        var bestD = Double.infinity
        for nbId in regions[smallId].neighbors where alive[nbId] {
            let d = labDistance2000(labOf[smallId], labOf[nbId])
            if d < bestD { bestD = d; bestId = nbId }
        }
        if bestId < 0 { continue }

        // Absorb: rewire neighbor sets and add areas. No pixel arrays touched.
        for nbId in regions[smallId].neighbors {
            if nbId == bestId { continue }
            regions[nbId].neighbors.remove(smallId)
            regions[nbId].neighbors.insert(bestId)
            regions[bestId].neighbors.insert(nbId)
        }
        regions[bestId].neighbors.remove(smallId)
        regions[bestId].area += regions[smallId].area
        absorbedInto[smallId] = bestId

        alive[smallId] = false
        regions[smallId].area = 0
        regions[smallId].neighbors = []

        if regions[bestId].area < minAreaPixels { smallQueue.append(bestId) }
    }

    compactAndRebuild(&seg, &regions, alive, absorbedInto, width, height)
}

/// Compact region ids (no gaps), rebuild pixel lists and bboxes in one pass,
/// and remap neighbor sets to the new ids.
private func compactAndRebuild(_ seg: inout SegmentResult, _ regions: inout [Region], _ alive: [Bool], _ absorbedInto: [Int], _ width: Int, _ height: Int) {
    var remap = [Int](repeating: -1, count: regions.count)
    var kept: [Region] = []
    for i in 0..<regions.count where alive[i] {
        remap[i] = kept.count
        kept.append(regions[i])
    }

    var finalTarget = [Int](repeating: -1, count: regions.count)
    for i in 0..<regions.count {
        var t = i
        while !alive[t] && absorbedInto[t] >= 0 { t = absorbedInto[t] }
        finalTarget[i] = alive[t] ? t : -1
    }

    var buckets: [[Int]] = .init(repeating: [], count: kept.count)
    var bboxes = kept.map { _ in (minX: width, minY: Int.max, maxX: -1, maxY: -1) }
    let regionIds = seg.regionIds
    for p in 0..<regionIds.count {
        let origId = Int(regionIds[p])
        let target = alive[origId] ? origId : finalTarget[origId]
        let id = target >= 0 ? remap[target] : -1
        if id < 0 {
            seg.regionIds[p] = 0
            buckets[0].append(p)
            continue
        }
        seg.regionIds[p] = UInt32(id)
        buckets[id].append(p)
        let x = p % width
        let y = p / width
        if x < bboxes[id].minX { bboxes[id].minX = x }
        if x > bboxes[id].maxX { bboxes[id].maxX = x }
        if y < bboxes[id].minY { bboxes[id].minY = y }
        if y > bboxes[id].maxY { bboxes[id].maxY = y }
    }

    for i in 0..<kept.count {
        var neighbors = Set<Int>()
        for nb in kept[i].neighbors {
            let mapped = remap[nb]
            if mapped >= 0 && mapped != i { neighbors.insert(mapped) }
        }
        kept[i].id = i
        kept[i].pixels = buckets[i]
        kept[i].area = buckets[i].count
        kept[i].minX = bboxes[i].minX
        kept[i].minY = bboxes[i].minY
        kept[i].maxX = bboxes[i].maxX
        kept[i].maxY = bboxes[i].maxY
        kept[i].neighbors = neighbors
    }
    seg.regions = kept
    _ = height
}

File: src/engine/segment.ts
/** Connected regions + small-region merging per SPEC Task 6.
 * Designed for real photos: the merge loop is pure bookkeeping (no pixel
 * array concatenation); pixel lists and bboxes are rebuilt in a single
 * O(pixels) pass afterwards. */
import { labDistance2000 } from './colorSpace'
import type { Lab, PaletteEntry } from './types'

/** Contrast-aware merging (SPEC v3.1 P0): a region smaller than minArea is
 * merged into a neighbor ONLY if their ΔE00 is below this threshold. Small
 * but contrast-bearing regions (lips, eyelets, lace, blush) survive instead
 * of being dissolved into their surroundings. */
export const MERGE_CONTRAST_THRESHOLD = 12

export interface Region {
  id: number
  colorIdx: number
  area: number
  pixels: number[]
  bbox: { minX: number; minY: number; maxX: number; maxY: number }
  neighbors: Set<number>
}

export interface SegmentResult {
  /** Region id per pixel (compacted ids after merge). */
  regionIds: Uint32Array
  regions: Region[]
}

const UNASSIGNED = 0xffffffff

/** Split a label map into 4-connected regions and record adjacency
 * from pixel pairs with differing labels (right/down pairs). */
export function buildRegions(
  labels: Uint16Array,
  width: number,
  height: number,
): SegmentResult {
  const n = width * height
  const regionIds = new Uint32Array(n).fill(UNASSIGNED)
  const pixelsLists: number[][] = []
  const neighborSets: Set<number>[] = []
  const queue = new Int32Array(n)

  for (let start = 0; start < n; start++) {
    if (regionIds[start] !== UNASSIGNED) continue
    const colorIdx = labels[start]
    const id = pixelsLists.length
    const pixels: number[] = []
    neighborSets.push(new Set<number>())
    let head = 0
    let tail = 0
    queue[tail++] = start
    regionIds[start] = id
    pixels.push(start)

    while (head < tail) {
      const p = queue[head++]
      const x = p % width
      const y = (p - x) / width

      // 4-connectivity: right, down, left, up.
      // Same-label unassigned neighbors are claimed; differing-label neighbors
      // record adjacency once both regions exist (the reverse pair is always
      // examined by the other region's BFS, so nothing is missed).
      const q4 = [x + 1 < width ? p + 1 : -1, y + 1 < height ? p + width : -1, x > 0 ? p - 1 : -1, y > 0 ? p - width : -1]
      for (const q of q4) {
        if (q < 0) continue
        if (labels[q] === colorIdx) {
          if (regionIds[q] === UNASSIGNED) {
            regionIds[q] = id
            queue[tail++] = q
            pixels.push(q)
          }
        } else if (regionIds[q] !== UNASSIGNED) {
          addAdjacency(neighborSets, id, regionIds[q] as number)
        }
      }
    }

    pixelsLists.push(pixels)
  }

  const regions: Region[] = pixelsLists.map((pixels, id) => ({
    id,
    colorIdx: labels[pixels[0]] as number,
    area: pixels.length,
    pixels,
    bbox: { minX: 0, minY: 0, maxX: 0, maxY: 0 },
    neighbors: neighborSets[id],
  }))
  recomputeBboxes(regions, regionIds, width, height)

  return { regionIds, regions }
}

function addAdjacency(neighborSets: Set<number>[], a: number, b: number): void {
  if (a === b) return
  neighborSets[a].add(b)
  neighborSets[b].add(a)
}

function recomputeBboxes(
  regions: Region[],
  regionIds: Uint32Array,
  width: number,
  height: number,
): void {
  for (const r of regions) {
    r.bbox = { minX: width, minY: height, maxX: -1, maxY: -1 }
  }
  for (let p = 0; p < regionIds.length; p++) {
    const r = regions[regionIds[p]]
    const x = p % width
    const y = (p - x) / width
    if (x < r.bbox.minX) r.bbox.minX = x
    if (x > r.bbox.maxX) r.bbox.maxX = x
    if (y < r.bbox.minY) r.bbox.minY = y
    if (y > r.bbox.maxY) r.bbox.maxY = y
  }
}

/** Merge every region smaller than minAreaPixels into its nearest neighbor
 * until none remain (isolated small regions are kept). Two merging regimes:
 *  - contrast-aware pass: contrast small regions (ΔE00 ≥ threshold with every
 *    neighbor) are KEPT — details like lips/lace survive;
 *  - everything else merges into the ΔE00-nearest alive neighbor.
 * Only area/neighbor bookkeeping happens here; pixels are rebuilt once. */
export function mergeSmallRegions(
  seg: SegmentResult,
  palette: PaletteEntry[],
  minAreaPixels: number,
  width: number,
  height: number,
): void {
  if (minAreaPixels <= 0) return
  const { regions } = seg
  const alive = regions.map(() => true)
  /** absorbedInto[a] = the region that finally absorbed a's pixels. */
  const absorbedInto = new Int32Array(regions.length).fill(-1)

  const smallQueue: number[] = []
  for (const r of regions) {
    if (r.area < minAreaPixels) smallQueue.push(r.id)
  }

  const labOf: Lab[] = regions.map((r) => palette[r.colorIdx].lab)
  const dE00 = (a: number, b: number): number => labDistance2000(labOf[a], labOf[b])

  let qi = 0
  while (qi < smallQueue.length) {
    const smallId = smallQueue[qi++]
    if (!alive[smallId]) continue
    const small = regions[smallId]
    if (small.area >= minAreaPixels) continue
    if (small.neighbors.size === 0) continue // isolated: keep as-is

    // Contrast guard: a small region that stands out against EVERY alive
    // neighbor is a detail, not noise — keep it.
    let maxNeighborD = -1
    for (const nbId of small.neighbors) {
      if (!alive[nbId]) continue
      const d = dE00(smallId, nbId)
      if (d > maxNeighborD) maxNeighborD = d
    }
    if (maxNeighborD < 0) continue // no alive neighbors: keep
    if (maxNeighborD >= MERGE_CONTRAST_THRESHOLD) continue // contrast detail: keep

    // Nearest alive neighbor by ΔE00 (perceptual) among the low-contrast
    // candidates — the merge target must itself be a low-contrast neighbor.
    let bestId = -1
    let bestD = Infinity
    for (const nbId of small.neighbors) {
      if (!alive[nbId]) continue
      const d = dE00(smallId, nbId)
      if (d < bestD) {
        bestD = d
        bestId = nbId
      }
    }
    if (bestId < 0) continue

    // Absorb: rewire neighbor sets and add areas. No pixel arrays touched.
    const into = regions[bestId]
    for (const nbId of small.neighbors) {
      if (nbId === bestId) continue
      regions[nbId].neighbors.delete(smallId)
      regions[nbId].neighbors.add(bestId)
      into.neighbors.add(nbId)
    }
    into.neighbors.delete(smallId)
    into.area += small.area
    // Track the absorption chain so compaction can route the small region's
    // pixels (and any pixels of regions absorbed into it earlier) to `into`.
    absorbedInto[smallId] = bestId

    alive[smallId] = false
    small.area = 0
    small.neighbors.clear()

    // Absorption can drag the target below minArea; re-evaluate it, but it
    // re-enters the merge only if it is still low-contrast vs its neighbors.
    if (into.area < minAreaPixels) smallQueue.push(bestId)
  }

  compactAndRebuild(seg, alive, absorbedInto, width, height)
}

/** Compact region ids (no gaps), rebuild pixel lists and bboxes in one pass,
 * and remap neighbor sets to the new ids. Pixels of absorbed regions are
 * routed to their final absorber via the absorption chain. */
function compactAndRebuild(
  seg: SegmentResult,
  alive: boolean[],
  absorbedInto: Int32Array,
  width: number,
  height: number,
): void {
  void height
  const remap = new Int32Array(seg.regions.length).fill(-1)
  const kept: Region[] = []
  for (let i = 0; i < seg.regions.length; i++) {
    if (!alive[i]) continue
    remap[i] = kept.length
    kept.push(seg.regions[i])
  }

  // Resolve each dead region's final absorber once (chains are short: each
  // hop moves to an alive-or-later-absorbed region, but we guard with a walk).
  const finalTarget = new Int32Array(seg.regions.length)
  for (let i = 0; i < seg.regions.length; i++) {
    let t = i
    while (!alive[t] && absorbedInto[t] >= 0) t = absorbedInto[t]
    finalTarget[i] = alive[t] ? t : -1
  }

  const buckets: number[][] = kept.map(() => [])
  const bboxes = kept.map(() => ({ minX: width, minY: Infinity, maxX: -1, maxY: -1 }))
  const { regionIds } = seg
  for (let p = 0; p < regionIds.length; p++) {
    const origId = regionIds[p]
    const target = alive[origId] ? origId : finalTarget[origId]
    const id = target >= 0 ? remap[target] : -1
    if (id < 0) {
      // Unreachable for well-formed input (every pixel belongs to a region
      // that stayed alive or was absorbed); guard against corrupt maps.
      regionIds[p] = 0
      buckets[0].push(p)
      continue
    }
    regionIds[p] = id
    buckets[id].push(p)
    const x = p % width
    const y = (p - x) / width
    const b = bboxes[id]
    if (x < b.minX) b.minX = x
    if (x > b.maxX) b.maxX = x
    if (y < b.minY) b.minY = y
    if (y > b.maxY) b.maxY = y
  }

  seg.regions = kept.map((r, i) => {
    const neighbors = new Set<number>()
    for (const nb of r.neighbors) {
      const mapped = remap[nb]
      if (mapped >= 0 && mapped !== i) neighbors.add(mapped)
    }
    r.id = i
    r.pixels = buckets[i]
    r.area = buckets[i].length
    r.bbox = bboxes[i]
    r.neighbors = neighbors
    return r
  })
}
